fuzz/cli.rs

```diff
@@ -0,0 +1,252 @@
+    if FUZZ_TARGETS.iter().find(|x| *x == target).is_none() {
+        panic!(
+            "Unknown fuzz target `{}`. Run `list-targets` command to see available fuzz targets.",
+            target
+        );
+    }
+    write_fuzz_target_source_file(fuzzer, target)?;
+    match fuzzer {
+        Fuzzer::Afl => run_afl(target),
+        Fuzzer::Honggfuzz => run_honggfuzz(target),
+        Fuzzer::Libfuzzer => run_libfuzzer(target),
+    }
+}
+
+/// Run one target fuzz test using AFL
+fn run_afl(_filter: &str) -> Result<(), Error> {
+    // AFL requires initial inputs, so leave it to future.
+    // General process:
+    // 1. cargo afl build (in fuzzer-afl directory)
+    // 2. cargo afl fuzz -i in -o out target/debug/fuzzer-afl
+    unimplemented!()
+}
+
+/// Run one target fuzz test using Honggfuzz
+fn run_honggfuzz(target: &str) -> Result<(), Error> {
+    let fuzzer = Fuzzer::Honggfuzz;
+
+    let mut rust_flags = env::var("RUSTFLAGS").unwrap_or_default();
+    rust_flags.push_str("-Z sanitizer=address");
+
+    let mut hfuzz_args = env::var("HFUZZ_RUN_ARGS").unwrap_or_default();
+    hfuzz_args.push_str("--exit_upon_crash --run_time 5");
+
+    let fuzzer_bin = Command::new("cargo")
+        .args(&["hfuzz", "run", target])
+        .env("RUSTFLAGS", &rust_flags)
+        .env("HFUZZ_RUN_ARGS", &hfuzz_args)
+        .current_dir(fuzzer.directory())
+        .spawn()
+        .context(format!("Failed to run {}", fuzzer))?
+        .wait()
+        .context(format!("Failed to wait {}", fuzzer))?;
+
+    if !fuzzer_bin.success() {
+        Err(format_err!(
+            "{} exited with code {:?}",
+            fuzzer,
+            fuzzer_bin.code()
+        ))?;
+    }
+
+    Ok(())
+}
+
+/// Run one target fuzz test using Libfuzzer
+fn run_libfuzzer(target: &str) -> Result<(), Error> {
+    let fuzzer = Fuzzer::Libfuzzer;
+
+    #[cfg(target_os = "macos")]
+    let target_platform = "x86_64-apple-darwin";
+    #[cfg(target_os = "linux")]
+    let target_platform = "x86_64-unknown-linux-gnu";
+    #[cfg(not(any(target_os = "linux", target_os = "macos")))]
+    panic!("libfuzzer-sys only supports Linux and macOS");
+
+    let mut rust_flags = env::var("RUSTFLAGS").unwrap_or_default();
+    rust_flags.push_str(
+        "--cfg fuzzing \
+         -C passes=sancov \
+         -C llvm-args=-sanitizer-coverage-level=4 \
+         -C llvm-args=-sanitizer-coverage-trace-pc-guard \
+         -C llvm-args=-sanitizer-coverage-trace-compares \
+         -C llvm-args=-sanitizer-coverage-trace-divs \
+         -C llvm-args=-sanitizer-coverage-trace-geps \
+         -C llvm-args=-sanitizer-coverage-prune-blocks=0 \
+         -C debug-assertions=on \
+         -C debuginfo=0 \
+         -C opt-level=3 \
+         -Z sanitizer=address",
+    );
+
+    let mut asan_options = env::var("ASAN_OPTIONS").unwrap_or_default();
+    asan_options.push_str(" detect_odr_violation=0");
+
+    let fuzzer_bin = Command::new("cargo")
```

Comment:
Why not `cargo-fuzz`? So we dont need to maintain magic rust flags.

Reply:
cargo-fuzz requires magic Cargo.toml and a magic directory. No way to be good organized.